pkg/pillar/containerd/containerd.go

```diff
@@ -69,77 +71,90 @@ const (
 var (
 	// default snapshotter used by containerd
 	defaultSnapshotter = "overlayfs"
-	ctrdCtx            context.Context
-	ctrdSystemCtx      context.Context
-	// CtrdClient is a handle to the current containerd client API
-	CtrdClient   *containerd.Client
-	contentStore content.Store
 )
 
-// InitContainerdClient initializes CtrdClient and ctrdCtx
-func InitContainerdClient() error {
-	log.Infof("InitContainerdClient")
-	var err error
-	if ctrdCtx == nil {
-		ctrdCtx = namespaces.WithNamespace(context.Background(), ctrdServicesNamespace)
-	}
-	if ctrdSystemCtx == nil {
-		ctrdSystemCtx = namespaces.WithNamespace(context.Background(), ctrdSystemServicesNamespace)
-	}
-	if CtrdClient == nil {
-		CtrdClient, err = containerd.New(ctrdSocket, containerd.WithDefaultRuntime(containerdRunTime))
-		if err != nil {
-			log.Errorf("InitContainerdClient: could not create containerd client. %v", err.Error())
-			return fmt.Errorf("initContainerdClient: could not create containerd client. %v", err.Error())
-		}
+var onceLock = &sync.Mutex{}
+var onceVal bool
+
+// once returns true the first time and then false
+func once() bool {
+	onceLock.Lock()
+	defer onceLock.Unlock()
+	if onceVal {
+		return false
+	} else {
+		onceVal = true
+		return true
 	}
-	if contentStore == nil {
-		contentStore = CtrdClient.ContentStore()
+}
+
+// Client is the handle we return to the caller
+type Client struct {
+	ctrdClient   *containerd.Client
+	contentStore content.Store
+}
+
+// NewContainerdClient returns a *Client
+// Callable from multiple go-routines.
+// createClient=false is used by tests where we just need the contexts etc
+func NewContainerdClient(createClient bool) (*Client, error) {
+	log.Infof("NewContainerdClient")
+	var err error
+	if once() {
```

Comment:
Why all this once() business instead of init() function?

Reply:
We could put what is left inside once() into an init() function.

Reply:
Done.